spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/SimpleDirectMessageListenerContainer.java

```diff
@@ -0,0 +1,218 @@
+	 */
+	public void setConsumersPerQueue(int consumersPerQueue) {
+		this.consumersPerQueue = consumersPerQueue;
+	}
+
+	/**
+	 * Set to false to requeue failed deliveries.
+	 * @param requeueRejected the requeue rejected.
+	 */
+	public void setRequeueRejected(boolean requeueRejected) {
+		this.requeueRejected = requeueRejected;
+	}
+
+	@Override
+	protected void doInitialize() throws Exception {
+		Assert.notNull(getQueueNames(), "queue(s) are required");
+	}
+
+	@Override
+	protected void doStart() throws Exception {
+		super.doStart();
+		if (this.taskExecutor == null) {
+			this.taskExecutor = new SimpleAsyncTaskExecutor((getBeanName() == null ? "container" : getBeanName()) + "-");
+		}
+		this.taskExecutor.execute(new Runnable() {
+
+			boolean initialized = false;
+
+			@Override
+			public void run() {
+				while (!initialized) {
+					try {
+						for (int i = 0; i < consumersPerQueue; i++) {
+							for (String queue : getQueueNames()) {
+								Connection connection = getConnectionFactory().createConnection();
+								Channel channel = connection.createChannel(false);
+								channel.basicQos(prefetch);
+								RabbitUtils.setPhysicalCloseRequired(true);
+								SimpleConsumer consumer = new SimpleConsumer(channel, queue);
+								consumers.add(consumer);
+								channel.basicConsume(queue, consumer);
+							}
+						}
+					}
+					catch (Exception e) {
+						logger.error("Error creating consumer", e);
+						doShutdown();
+						try {
+							Thread.sleep(5000);
+						}
+						catch (InterruptedException e1) {
+							Thread.currentThread().interrupt();
+						}
+						continue;
+					}
+					initialized = true;
+				}
+			}
+
+		});
+	}
+
+	@Override
+	protected void doShutdown() {
+		Assert.state(this.taskExecutor != null, "Cannot shut down if not initialized");
+		for (DefaultConsumer consumer : this.consumers) {
+			try {
+				consumer.getChannel().basicCancel(consumer.getConsumerTag());
+				consumer.getChannel().close();
+			}
+			catch (IOException e) {
+				logger.error("Cancel Error", e);
+			}
+			catch (TimeoutException e) {
+				logger.error("Timeout Error on channel close", e);
+			}
+		}
+		this.consumers.clear();
+	}
+
+	private final class SimpleConsumer extends DefaultConsumer {
+
+		private final String queue;
+
+		public SimpleConsumer(Channel channel, String queue) {
+			super(channel);
+			this.queue = queue;
+		}
+
+		@Override
+		public void handleDelivery(String consumerTag, Envelope envelope,
+				BasicProperties properties, byte[] body) throws IOException {
+			MessageProperties messageProperties = messagePropertiesConverter.toMessageProperties(
+					properties, envelope, "UTF-8");
+			messageProperties.setMessageCount(0);
+			messageProperties.setConsumerTag(consumerTag);
+			messageProperties.setConsumerQueue(queue);
+			Message message = new Message(body, messageProperties);
+			try {
+				invokeListener(message, getChannel());
+				getChannel().basicAck(envelope.getDeliveryTag(), false);
```

Comment:
I wonder is it idempotent to `basicAck()` from my target listener (e.g. over `AmqpHeaders.CHANNEL` in the SI app) and then here one more time...